### PR TITLE
Allow more than one layer of allOf inheritance

### LINKED ISSUE #422: Inherited fields not generated for more than one level of "allOf" inheritance
### General information

Schemas using `allOf` inheritance do not generate form fields for more than one level of inheritance.

- json-editor version:1.4.0-beta.0

### Expected behavior

Should generate form fields for all required fields in inheritance hierarchy.

### Actual behavior

Does not generate grandparent field. Online editor behaves inconsistently thereafter.

### Steps to reproduce the behavior

1. Open (https://json-editor.github.io/json-editor/) paste sample schema (below) into Schema text box and click `Update Schema`. fields declared directly in the child object schema are correctly generated, as are those in the parent, but not those in the grandparent. A validation error is shown for the missing grandparent field.
2. Provide a value for the missing field by updating the JSON output. The validation error disappears, and the field appears in the form but with a type selector.
3. If the field is then removed from the JSON output, the form field disappears but the validation error does not reappear.
4. Repeat step 1 and click "Direct Link". The grandparent field still does not appear but neither does the validation error. Also the parentField (`numLegs`) now has a type selector.
5. Remove the numLegs field from the JSON output. The form field disappears but there is no validation error.

```json
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "definitions": {
        "animal": {
            "type": "object",
            "properties": {
                "id": {
                    "type": "number"
                }
            },
            "required": [
                "id"
            ]
        },
        "landAnimal": {
            "allOf": [
                {
                    "$ref": "#/definitions/animal"
                }
            ],
            "type": "object",
            "title": "Land Animal",
            "properties": {
                "numLegs": {
                    "type": "integer"
                }
            },
            "required": [
                "numLegs"
            ]
        }
    },
    "allOf": [
        {
            "$ref": "#/definitions/landAnimal"
        }
    ],
    "type": "object",
    "title": "Cat",
    "properties":{
        "breed":{"type":"string"}
    },
    "required":["breed"]
}
```

### PR DESCRIPTION
**Pull request by sirockin**
Fixes #422.

The changes required were fortunately fairly simple.

I've added a stub e2e test which simply checks all `required` fields in a two-layer hierarchy are added.

I scratched my head for about an hour wondering why four other e2e tests were failing and eventually worked out they were failing in the main repo too, but were being ignored by Travis because of the `@optional` tag. So I added some text to `./testing/README.md` to point this out to future contributors.

### CODE REVIEW

File: src/core.js [hunk 1]
```diff
@@ -440,7 +440,7 @@ JSONEditor.prototype = {
       callback();
     }
   },
-  expandRefs: function(schema) {
+  expandRefs: function(schema, recurseAllOf) {
```
Comment (pmk65) [inline src/core.js]: @sirockin Have you checked if any of the other instances of `expandRef` benefits from adding the extra Boolean parameter? (Note: the line numbers might not match the version you are using)

**\json-editor\src\core.js           3 matches**
[56:25]	      var schema = self.expandRefs(self.schema);
[443:3]	  expandRefs: function(schema) {
[545:17]	    return this.expandRefs(extended);

**\json-editor\src\editors\array.js           3 matches**
[181:35]	        var tmp = this.jsoneditor.expandRefs(this.schema.items);
[220:30]	    schema = this.jsoneditor.expandRefs(schema);
[234:30]	    schema = this.jsoneditor.expandRefs(schema);

**\json-editor\src\editors\multiple.js           1 matches**
[87:32]	      schema = self.jsoneditor.expandRefs(schema);

**\json-editor\src\editors\multiselect.js           1 matches**
[34:40]	        items_schema = this.jsoneditor.expandRefs(this.schema.items || {}),

**\json-editor\src\editors\table.js           1 matches**
[22:39]	    var item_schema = this.jsoneditor.expandRefs(this.schema.items || {});

**\json-editor\src\validator.js           1 matches**
[20:41]	    schema = $extend({},this.jsoneditor.expandRefs(schema));

Comment (sirockin) [reply src/core.js]: Not yet, no. I haven't completely got my head around the interactions going on here so I'm nervous about breaking any existing functionality. I'm also on a bit of a deadline with the work I'm doing right now. 

How confident are you that the e2e tests cover existing functionality? It's always reassuring to see tests pass after changing code :-)

I have a feeling I'll be working on this again so will keep in mind your comments and look at the other invocations next time I dig in there.

Comment (pmk65) [reply src/core.js]: I can't see any problems with applying your patch to all instances of `expandRefs`. If the functionality is currently broken. 
And `oneOf`, `anyOf` and `allOf` all "share" the same functionality, so the other 2 probably have same problems.
As for the tests, then I don't think they are 100%. We discover issues not covered by the tests from time to time. 😄

**Comment by schmunk42** [comment]
Thank you!

**Pull request merged**